sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/QueueingBeamFnDataClient.java

```diff
@@ -141,6 +138,11 @@ public void drainAndBlock() throws Exception {
     return this.mainClient.send(apiServiceDescriptor, outputLocation, coder);
   }
 
+  /** Resets this object so that it may be reused. */
+  public void reset() {
+    inboundDataClients.clear();
```

Comment:
Is there a need to check if InboundDataClients are done?

Reply:
I don't have better opinion but sounds like recreating this object make sense?

Reply:
I looked into that some more but it isn't easy since the queue reference has made it's way into various places as part of the cached BundleProcessor.  I modified things so that the created QueueingFnDataReceiver are bound to an explicit queue instead of the QueueingBeamFnDataClient itself.  This means that in reset we can just create a new queue and the identified race doesn't matter.  Clients that have been failed but maybe not yet done will add to the orphaned queue (or if the queue is full they will abort once they notice they are failed).